File: 94.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        // 递归
        // List<Integer> ans = new ArrayList<>();
        // inOrder(ans, root);
        // return ans;

        // 迭代
        // List<Integer> ans = new ArrayList<>();
        // Deque<TreeNode> stack = new LinkedList<>();
        // while(root != null || !stack.isEmpty()) {
        //     while(root != null) {
        //         stack.push(root);
        //         root = root.left;
        //     }
        //     root = stack.pop();
        //     ans.add(root.val);
        //     root = root.right;
        // }
        // return ans;

        //Morris
        List<Integer> ans = new ArrayList<>();
        TreeNode pre = null;
        while(root != null) {
            if(root.left == null) {
                ans.add(root.val);
                root = root.right;
            } else {
                pre = root.left;
                while(pre.right != null && pre.right != root) pre = pre.right;
                if(pre.right == null) {
                    pre.right = root;
                    root = root.left;
                } else if(pre.right == root) {
                    ans.add(root.val);
                    pre.right = null;
                    root = root.right;
                }
            }
        }
        return ans;
    }

    // public void inOrder(List<Integer> ans, TreeNode root) {
    //     if(root == null) return ;
    //     inOrder(ans, root.left);
    //     ans.add(root.val);
    //     inOrder(ans, root.right);
    // }
}

File: 94.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func inorderTraversal(root *TreeNode) []int {
	//递归
	// res := []int{}
	// inorder(root, &res)
	// return res

	//迭代
	stack := []*TreeNode{}
	res := []int{}
	for root != nil || len(stack) > 0 {
		for root != nil {
			stack = append(stack, root)
			root = root.Left
		}
		root = stack[len(stack)-1]
		res = append(res, root.Val)
		stack = stack[:len(stack)-1]
		root = root.Right
	}
	return res
}

// func inorder(node *TreeNode, res *[]int) {
//     if node == nil {
//         return
//     }
//     inorder(node.Left, res)
//     *res = append(*res, node.Val)
//     inorder(node.Right, res)
// }